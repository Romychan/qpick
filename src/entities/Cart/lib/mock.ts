// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
