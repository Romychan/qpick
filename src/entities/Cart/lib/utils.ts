// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
