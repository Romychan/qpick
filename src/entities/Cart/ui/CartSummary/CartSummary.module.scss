@import '~/app/assets/styles/_typography.scss';

.container {
  position: relative;
  padding: 0px;
  width: 100%;
  border: 1px solid var(--gray-200);
  border-radius: var(--radius-default);
  justify-self: flex-start;
}

.information {
  display: flex;
  justify-content: space-between;
  padding: 1.2rem;
}

.text {
  @include text-sm;
  @include text-semibold;
  text-transform: uppercase;
  color: var(--gray-1000);
}

.actions {
  width: 100%;

  & button {
    width: 100%;
  }
}
