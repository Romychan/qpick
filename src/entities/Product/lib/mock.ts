import { Product } from '../model/types';

/** List of mock products */
export const MOCK_PRODUCTS: Product[] = [
  {
    id: 1,
    title: 'Apple BYZ S852I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 2927,
    oldPrice: 3527,
    description:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    category: 'Наушники',
  },
  {
    id: 2,
    title: 'Apple AirPods Pro 3',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Наушники',
  },
  {
    id: 3,
    title: 'Apple EarPods',
    rate: 4.5,
    image: '/images/catalog/headphone-3.png',
    price: 2327,
    oldPrice: null,
    description:
      'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    category: 'Наушники',
  },
  {
    id: 4,
    title: 'Apple BYZ S123I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 1500,
    oldPrice: 1950,
    description:
      'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
    category: 'Наушники',
  },
  {
    id: 5,
    title: 'Apple AirPods Pro 2',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    category: 'Наушники',
  },
  {
    id: 6,
    title: 'Apple EarPods 2',
    rate: 3.8,
    image: '/images/catalog/headphone-3.png',
    price: 1200,
    oldPrice: 2000,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Наушники',
  },
  {
    id: 7,
    title: 'Apple AirPods',
    rate: 4.7,
    image: '/images/catalog/headphone-4.png',
    price: 9527,
    oldPrice: null,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Беспроводные наушники',
  },
  {
    id: 8,
    title: 'GERLAX GH-04',
    rate: 4.7,
    image: '/images/catalog/headphone-5.png',
    price: 6527,
    oldPrice: null,
    description:
      'Sed quis sem vitae justo feugiat tincidunt. Maecenas ut leo nisl. Praesent finibus ultricies ex et tristique.',
    category: 'Беспроводные наушники',
  },
  {
    id: 9,
    title: 'BOROFONE BO4',
    rate: 4.7,
    image: '/images/catalog/headphone-6.png',
    price: 7527,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Беспроводные наушники',
  },
];
