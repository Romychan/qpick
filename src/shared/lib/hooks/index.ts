export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
