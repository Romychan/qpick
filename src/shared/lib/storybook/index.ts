export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
