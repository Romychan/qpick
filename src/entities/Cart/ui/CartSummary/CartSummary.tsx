import { ReactNode, useMemo } from 'react';

import { Card } from '~/shared/ui/Card';
import { formatPrice } from '~/shared/lib/utils';

import { CartItem } from '../../model/types';

import styles from './CartSummary.module.scss';

interface ICartSummaryProps {
  /** The list with the product in the cart */
  cart: CartItem[];
  /** The slot of the component for interactive elements */
  actions: ReactNode;
}

/** This component is used to display information about the total price of products in the cart */
export const CartSummary = ({ cart, actions }: ICartSummaryProps) => {
  const totalPrice = useMemo(
    () => cart.reduce((acc, cur) => acc + cur.product.price * cur.quantity, 0),
    [cart],
  );

  return (
    <Card className={styles.container}>
      <div className={styles.information}>
        <span className={styles.text}>Итого</span>
        <span className={styles.text} data-testid="checkout-total-price">
          ₽ {formatPrice(totalPrice, 'decimal') || 0}
        </span>
      </div>

      <div className={styles.actions}>{actions}</div>
    </Card>
  );
};
