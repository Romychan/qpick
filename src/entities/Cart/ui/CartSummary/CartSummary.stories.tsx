import type { Meta, StoryObj } from '@storybook/react';

import {
  withRouterDecorator,
  withWidthDecorator,
} from '~/shared/lib/storybook';
import { Button, Default as DefaultButton } from '~/shared/ui/Button';

import { MOCK_CART } from '../../lib/mock';

import { CartSummary } from './CartSummary';

const meta = {
  title: 'Entities/Cart/CartSummary',
  component: CartSummary,
  parameters: {
    layout: 'centered',
    width: '350px',
  },
  args: {
    cart: MOCK_CART,
  },
  decorators: [withRouterDecorator, withWidthDecorator],
  tags: ['autodocs'],
} satisfies Meta<typeof CartSummary>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    actions: <Button {...DefaultButton.args}>Оформить</Button>,
  },
};
