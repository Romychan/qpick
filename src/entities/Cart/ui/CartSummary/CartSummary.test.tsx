import { render, screen } from '@testing-library/react';

import { MOCK_CART } from '../../lib/mock';

import { CartSummary } from './CartSummary';

describe('CartSummary', () => {
  it('should render correctly', () => {
    render(<CartSummary cart={MOCK_CART} actions={<button>Actions</button>} />);

    const totalPriceElement = screen.getByTestId('checkout-total-price');
    const actionsElement = screen.getByRole('button');

    expect(totalPriceElement).toBeInTheDocument();
    expect(totalPriceElement).toHaveTextContent(/2 400/i);
    expect(actionsElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const container = render(
      <CartSummary cart={MOCK_CART} actions={<button>Actions</button>} />,
    );

    expect(container).toMatchSnapshot();
  });
});
