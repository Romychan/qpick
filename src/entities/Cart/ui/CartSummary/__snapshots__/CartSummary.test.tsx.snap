// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartSummary > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_1d0052 _container_03b5b0"
      >
        <div
          class="_information_03b5b0"
        >
          <span
            class="_text_03b5b0"
          >
            Итого
          </span>
          <span
            class="_text_03b5b0"
            data-testid="checkout-total-price"
          >
            ₽ 
            2 400
          </span>
        </div>
        <div
          class="_actions_03b5b0"
        >
          <button>
            Actions
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_1d0052 _container_03b5b0"
    >
      <div
        class="_information_03b5b0"
      >
        <span
          class="_text_03b5b0"
        >
          Итого
        </span>
        <span
          class="_text_03b5b0"
          data-testid="checkout-total-price"
        >
          ₽ 
          2 400
        </span>
      </div>
      <div
        class="_actions_03b5b0"
      >
        <button>
          Actions
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
